feat: add keep_dropped_targets limit to CRDs

slashpai opened the pull request:
Fixes #5875

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._



## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

<!-- 
Your release note should be written in clear and straightforward sentences. Most often, users aren't familiar with
the technical details of your PR, so consider what they need to know when you write your release note.

Some brief examples of release notes:
- Add metadataConfig field to the Prometheus CRD for configuring how remote-write sends metadata information.
- Generate correct scraping configuration for Probes with empty or unset module parameter.
-->

```release-note
feat: add keep_dropped_targets limit to CRDs
```

ArthurSens commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -472,6 +472,16 @@ type CommonPrometheusFields struct {
+	// When defined, enforcedKeepDroppedTargets specifies a global limit on the number of targets
+	// dropped by relabeling that will be kept in memory. The value overrides
+	// any `spec.keepDroppedTargets` set by
+	// ServiceMonitor, PodMonitor, Probe objects unless `spec.keepDroppedTargets` is
```
```suggestion
	// ServiceMonitor, PodMonitor, Probe and ScrapeConfig objects unless `spec.keepDroppedTargets` is
```

ArthurSens replied:
Oh, I see that you've copied from the fields above... We can keep it consistent or update the other comments to include ScrapeConfig as well, up to you :)

slashpai replied:
Ya. I thought to create separate PR to update comments for all. Ok to add in this one too if needed. Just kept same way like other limits in this PR

ArthurSens commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -5510,6 +5521,263 @@ scrape_configs:
 }
 
+func TestKeepDroppedTargets(t *testing.T) {
+	expectNoKeepDroppedTargets := `global:
```
any special reason to not use golden files here?

slashpai replied:
updated

ArthurSens commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -5510,6 +5521,263 @@ scrape_configs:
+		{
+			version:                    "v2.47.0",
+			enforcedKeepDroppedTargets: 1000,
+			keepDroppedTargets:         500,
+			expected:                   fmt.Sprintf(expectKeepDroppedTargets, 500),
```
Ok, I can see that fmt.Sprintf makes it difficult to use golden files.

I'd be interested to know what @nicolastakashi thinks about this situation.

simonpasquier replied:
I'd still use golden files even if it means having more files in the testdata/ directory.

slashpai replied:
I will update that. Other limit tests were not using golden file I had used one of them to create this one.

slashpai replied:
updated

ArthurSens reviewed (approved):
Just some nits about comments and testing... definitely non-blocking

simonpasquier commented on pkg/apis/monitoring/v1/podmonitor_types.go:
```diff
@@ -74,6 +74,12 @@ type PodMonitorSpec struct {
 	LabelValueLengthLimit *uint64 `json:"labelValueLengthLimit,omitempty"`
+	// Per-scrape limit on the number of targets dropped by relabeling
+	// that will be kept in memory. 0 means no limit.
+	// Only valid in Prometheus versions 2.47.0 and newer.
```
```suggestion
	//
	// It requires Prometheus >= v2.47.0.
```

simonpasquier commented on pkg/apis/monitoring/v1/probe_types.go:
```diff
@@ -96,6 +96,12 @@ type ProbeSpec struct {
 	LabelValueLengthLimit *uint64 `json:"labelValueLengthLimit,omitempty"`
+	// Per-scrape limit on the number of targets dropped by relabeling
+	// that will be kept in memory. 0 means no limit.
+	// Only valid in Prometheus versions 2.47.0 and newer.
```
same here

simonpasquier commented on pkg/apis/monitoring/v1/prometheus_types.go:
```diff
@@ -580,6 +590,12 @@ type CommonPrometheusFields struct {
 	LabelValueLengthLimit *uint64 `json:"labelValueLengthLimit,omitempty"`
+	// Per-scrape limit on the number of targets dropped by relabeling
+	// that will be kept in memory. 0 means no limit.
+	// Only valid in Prometheus versions 2.47.0 and newer.
```
same here

simonpasquier commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -5510,6 +5521,263 @@ scrape_configs:
+
+	for _, tc := range []struct {
+		version                    string
+		enforcedKeepDroppedTargets int
```
any reason not using pointers here? It would simplify the setup/readability of the test.

slashpai replied:
I think other limit test were using int (before making them as pointer I types I believe). Since I had used one of those test for creating this one had kept the same, I am updating the test

slashpai replied:
updated

slashpai commented:
@simonpasquier updated with changes

slashpai commented on pkg/prometheus/promcfg_test.go:
```diff
@@ -5510,6 +5521,120 @@ scrape_configs:
+
+			if tc.enforcedKeepDroppedTargets != nil && *tc.enforcedKeepDroppedTargets >= 0 {
+				p.Spec.EnforcedKeepDroppedTargets = ptr.To(uint64(*tc.enforcedKeepDroppedTargets))
+			}
```
fixed

slashpai commented:
Fixed the test, I dont remember why int was used for limits in tests in past. I updated to use uint64 since thats the type anyways. Also removed redundant tests. I will create separate PR for other limit tests.
